microcosm_elasticsearch/store.py

```diff
@@ -186,3 +187,65 @@ def delete(self, identifier):
             using=self.elasticsearch_client,
         )
         return True
+
+    def _add_id_to_instance(self, instance):
+        """
+        Defines an ID for an instance if not previously defined
+
+        Set the internal _id equal to the instance ID
+
+        """
+        if instance.id is None:
+            instance.id = self.new_object_id()
+
+        instance._id = instance.id
+        return instance
+
+    def _add_op_type(self, instance, op_type):
+        """
+        Decorates an instance with a bulk operation type
+
+        """
+        instance["_op_type"] = op_type
+        return instance
+
+    def _batch_bulk(self, actions, batch_size):
+        """
+        Breaks list of actions into batches
+
+        """
+        num_actions = len(actions)
+        for offset in range(0, num_actions, batch_size):
+            yield actions[offset:min(offset + batch_size, num_actions)]
+
+    @translate_elasticsearch_errors
+    def bulk(self, actions, batch_size):
+        """
+        Bulk index entities
+
+        actions: list of tuples of (action, instance) to be included in the bulk
+        batch_size: number of records for each bulk call
+
+        All errors and exceptions are suppressed and are returned in the response report
+
+        """
+        actions = [
+            self._add_op_type(
+                instance=self._add_id_to_instance(instance).to_dict(True),
+                op_type=action,
+            )
+            for action, instance in actions
+        ]
+
+        return [
+            bulk(
+                client=self.elasticsearch_client,
+                actions=actions_batch,
+                index=self.index._name,
+                raise_on_exception=False,
+                raise_on_error=False,
```

Comment:
without knowing too much about details here, are we sure we want to set `raise_on_XXX` to False here by default rather than set to True and let application code decide if to catch and swallow? sounds like something that can swallow alot of silent errors we won't know about

Reply:
So it doesn't get complete swallowed, it is returned by the API with a list of items that failed (a bulk operation report). If we don't set these to False the entire bulk operation fails. This is especially problematic with the deprecated fields, if we attempt to delete them and they don't exist (not really an issue) it raises a `not_found` error which we don't care about. We can then take this bulk operation failed report to better inform what to do after the bulk calls happen.